### PR TITLE
Implement ISSN-based source validation system

### PR DESCRIPTION
**Pull request by muthukumaranR**
## Summary
Replaces fuzzy title matching with precise ISSN-based source validation for improved accuracy and maintainability. This change eliminates false positives/negatives from title variations while providing a more robust validation system.

## Key Changes
- **ISSN-based validation**: Uses CrossRef ISSN data instead of title matching
- **Flexible whitelist formats**: Supports flat lists, categorized maps, and nested structures
- **ISSN normalization**: Standardizes format and validates structure
- **ArXiv support**: Configurable bypass for preprint repositories
- **Enhanced debugging**: Detailed logging with journal names, DOIs, and URLs
- **Test improvements**: Fixed attribute references and moved to proper directory

## Technical Implementation
- Extracts ISSNs from CrossRef `ISSN` and `issn-type` fields
- Normalizes to `NNNN-NNNN` format with validation
- Supports multiple whitelist JSON formats for flexibility
- Maintains backward compatibility with existing search structures
- Adds `matched_issn` field for debugging and transparency

## Breaking Changes
- Default whitelist path: `pubs_whitelist.json` → `issn_whitelist.json`
- Schema change: `ValidationResult.journal_info` → `ValidationResult.source_info`
- Whitelist format: title-based → ISSN-based

## Testing
- [x] Created test suite with 26 tests (100% pass rate)
- [x] Covers ISSN normalization, whitelist loading, DOI extraction
- [x] Tests CrossRef integration, validation logic, end-to-end workflow
- [x] Includes edge cases, error handling, and configuration validation
- [x] Updated example script with proper attribute references
- [x] Verified backward compatibility with existing search results

## Migration Path
1. Create new `issn_whitelist.json` with ISSN list/categories
2. Update any code referencing `journal_info` to `source_info`
3. Optional: Enable `allow_arxiv: true` for preprint support

## Commits
- Replace title-based validation with ISSN matching
- Fix example script attribute references
- Add test suite for ISSN source validation
- Add ISSN whitelist documentation and mapping
- Add ISSN whitelist builder script

### CODE REVIEW

**Comment by NISH1001** [comment]
@muthukumaranR can you put the usage example code snippet as well. Thanks.

**Comment by NISH1001** [comment]
@muthukumaranR Is this an active PR?

File: akd/tools/source_validator.py [hunk 1]
```diff
@@ -384,97 +400,84 @@ async def _validate_single_result(
...
-        Returns:
-            Validation results
-        """
-        # Update whitelist path if provided
         if params.whitelist_file_path:
             self.config.whitelist_file_path = params.whitelist_file_path
-            self._whitelist = self._load_whitelist()
-
-        validated_results = []
+            self._load_whitelist()
```
Comment (NISH1001) [inline akd/tools/source_validator.py:438]: This has nasty side effect. We don't want to change the config at runtime based on param `self.config.whitelist_file_path = params.whitelist_file_path`

Intead maybe just have whitelist preloaded and not called from `_arun(...)`. 

Modifying self.config during execution breaks thread safety in general...since it also setting/building up these attributes once.

```
            self._allowed_issn_set = allowed_issn
            self._issn_to_category = issn_to_category
```

Since we're setting these params in `load_whitelist(...)`, so dynamically changing the config might be bad idea. Maybe what we can do is `self._load_whitelist(<path_str>)` is a better way.

File: akd/tools/source_validator.py [hunk 2]
```diff
@@ -136,8 +154,10 @@ def __init__(
         config.debug = debug
         super().__init__(config, debug)
 
-        # Load whitelist on initialization
-        self._whitelist = self._load_whitelist()
+        # Load whitelist on initialization (ISSN-based)
+        self._allowed_issn_set: Set[str] = set()
+        self._issn_to_category: Dict[str, str] = {}
+        self._load_whitelist()
```
Comment (NISH1001) [inline akd/tools/source_validator.py:160]: Delegate this to maybe `_post_init()`?

```
def _post_init(self):
  super()._post_init()
  self._load_whitelist()
```

File: akd/tools/source_validator.py [hunk 3]
```diff
@@ -21,16 +23,15 @@
...
 class SourceInfo(BaseModel):
     """Schema for source information from CrossRef API."""
 
-    title: str = Field(..., description="Source title")
+    title: str = Field(..., description="Source title (journal/proceedings)")
     publisher: Optional[str] = Field(None, description="Publisher name")
-    issn: Optional[List[str]] = Field(None, description="List of ISSNs")
+    issn: List[str] = Field(
+        default_factory=list, description="List of normalized ISSNs"
+    )
```
Comment (NISH1001) [inline akd/tools/source_validator.py:34]: I see issn changed from list[str] to optional[list[str]]. Not sure what are the consequences. Are we guaranteeing some issn?

File: akd/tools/source_validator.py [hunk 4]
```diff
@@ -384,97 +400,84 @@ async def _validate_single_result(
...
-        # Create aiohttp session with semaphore for concurrent requests
         connector = aiohttp.TCPConnector(limit=self.config.max_concurrent_requests)
         timeout = aiohttp.ClientTimeout(total=self.config.timeout_seconds)
 
         async with aiohttp.ClientSession(
             connector=connector,
             timeout=timeout,
         ) as session:
-            # Process results concurrently but with limited concurrency
             import asyncio
```
Comment (NISH1001) [inline akd/tools/source_validator.py:448]: Remove the import inside methods

File: akd/tools/source_validator.py [hunk 5]
```diff
@@ -152,67 +172,89 @@ def __init__(
...
+            return None
+        candidate = re.sub(r"[^0-9xX]", "", issn).upper()
+        return f"{candidate[:4]}-{candidate[4:]}" if len(candidate) == 8 else None
 
+    def _extract_doi_from_url(self, url: str) -> Optional[str]:
         for pattern in self._doi_patterns:
-            matches = re.findall(pattern, url_str, re.IGNORECASE)
+            matches = re.findall(pattern, str(url).strip(), re.IGNORECASE)
             if matches:
                 doi = matches[0] if isinstance(matches[0], str) else matches[0][0]
```
Comment (NISH1001) [inline akd/tools/source_validator.py:247]: Possibly cause indexing error for match?

File: akd/tools/source_validator.py [hunk 6]
```diff
@@ -152,67 +172,89 @@ def __init__(
...
-        Returns:
-            Extracted DOI or None if not found
-        """
-        url_str = str(url).strip()
+    @staticmethod
+    def _normalize_issn(issn: str) -> Optional[str]:
+        if not issn:
+            return None
+        candidate = re.sub(r"[^0-9xX]", "", issn).upper()
+        return f"{candidate[:4]}-{candidate[4:]}" if len(candidate) == 8 else None
```
Comment (NISH1001) [inline akd/tools/source_validator.py:241]: Should we add digit check as well?
> _"To confirm the check digit, calculate the sum of all eight digits of the ISSN multiplied by their position in the number, counting from the right"_

This might be handy in case the upstream is actually LLM generated ISSN?

File: akd/tools/source_validator.py [hunk 7]
```diff
@@ -384,97 +400,84 @@ async def _validate_single_result(
...
         ) as session:
-            # Process results concurrently but with limited concurrency
             import asyncio
 
             semaphore = asyncio.Semaphore(self.config.max_concurrent_requests)
 
-            async def validate_with_semaphore(
-                result: Any,
-            ) -> ValidationResult:
+            async def validate_with_semaphore(result: Any) -> ValidationResult:
```
Comment (NISH1001) [inline akd/tools/source_validator.py:451]: Is there a reason for using `Any`?

**Review by NISH1001 — CHANGES_REQUESTED** [review]
